feat: add dnf charm library

NucciTheBoss opened the pull request:
## Description

This pull request aims to add a dnf charm library offering to the operator_libs_linux collection of charm libraries. The dnf library will be used for charms that need to support both Ubuntu and RPM-based distributions (CentOS, Alma, etc.), as well as charms just intended to be deployed to just RPM-based Juju units. It comes with the standard bells and whistles needed to interact with dnf from a simple Python interface.

## What still needs to be done

Documentation and registering the library on Charmhub. This pull request is mostly focused on just adding the initial code for a dnf charm library; do not want too many themes in the pull request review.

## Notes

* I updated the version of the `setup-lxd` GitHub Action to 0.1.0 instead of just using a commit hash. I also set the channel of LXD to 5.9/stable
* I am using the [cleantest](https://nuccitheboss.github.io/cleantest/) library to provide the test instances for the dnf charm library. This way the library can be tested inside a RPM-based environment on GitHub Actions.
* I added a separate tox env named `integration-but-with-cleantest` so that I did not mess with how the current integration tests are done. I added an extra step in the CI job for the new environment.

jnsgruk commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+
+
+class _MetaDNF(type):
+    """Metaclass for `dnf`.
```
This feels like an over-complication to me, consider if this is an approach that brings you significant value over a more simple approach

NucciTheBoss replied:
I used this approach as a forcing function to inform the end-user that the charm base they are using does not support DNF. I used a metaclass so that this check/error being raised would be done at the dnf class definition rather than needing to have a conditional be evaluated every time the end-user invokes a dnf command.

I also used the metaclass because I wanted the dnf class to have a `__getitem__`-like syntax for pulling up information about packages, and I wanted to have a class property for retrieving the version of the dnf binary.

NucciTheBoss replied:
Removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

jnsgruk commented on tests/integration/test_dnf.py:
```diff
@@ -0,0 +1,170 @@
+import pathlib
+import unittest
+
+from cleantest.control import Env
+from cleantest.data import File
+from cleantest.provider import LXDArchon, lxd
```
I'm a little nervous of introducing this **right now**. Clever though it is, cleantest is relatively young, and also not used anywhere else in this project.

Given that all of the useful tests in this file use AlmaLinux 9, perhaps we could just use the AlmaLinux [cloud images for LXD](https://uk.lxd.images.canonical.com/), and take a similar approach to the existing functional tests that run on Ubuntu?

jnsgruk replied:
Just to clarify, by "useful tests" I mean that we probably don't need to bother firing up a Ubuntu LXD instance just to validate that `dnf` doesn't work - I'd save yourself the effort/code!

jnsgruk replied:
I think at this point it's just a little **too new** and unsupported for me to accept it. Remember that this repo supports libraries used by the whole charm ecosystem, and we need to ensure that the tests can be run reliably, without relying on your goodwill to maintain `cleantest` :)

I think for now, we should mimic the behaviour in the original tests, please. Happy for you to open an issue and we can investigate how to simplify the tox file (and multiple environments) in a separate thread - and perhaps one day that *could* involve cleantest - but only after a more thorough evaluation.

@benhoyt any parting words here on this topic?

NucciTheBoss replied:
That's a fair assessment. I already opened this issue [here](https://github.com/canonical/operator-libs-linux/issues/76) that pertains to improving the tests on this repo. It circles back to a discussion that Ben and I had somewhere in this thread :sweat_smile:

I will make the dnf tests inline with the current integration tests. Perhaps we can have a more in-depth discussion about cleantest at the next sprint or sometime sooner :smile:

benhoyt replied:
Yeah, I think it's wise to at least do a separate assessment. Though I don't mind using a simple library/repo like this as a bit of a test bed ... but let's at least leave to a future discussion/PR.

jnsgruk reviewed (changes requested):
A lot of this is just nits on style, but I do think in places we're over-complicating things and we probably need to consider if the added complexity is bringing much value over a more simplistic approach.

Remember that in the life of the charm, these methods are likely to be used only during `install` and `upgrade-charm`, which represents a very small portion of the charm's life, and never concurrently.

benhoyt reviewed (changes requested):
I think a lot of this is overly complex, and I'm worried about the "available" cache and the side effects done at import time. I've submitted a lot of comments to this effect. I hope they're constructive!

I think we should look at some use cases to determine what we actually need. And ask whether the cache is needed, or can we use something like `dnf info` instead.

benhoyt commented:
Oh sorry, I spent a bunch of time reviewing this and just submitted my comments, but didn't see that @jnsgruk just had too. Looks like we have many of the same exact concerns. Sorry about the duplication!

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+
+
+class PackageInfo:
+    """Dataclass representing DNF package information."""
```
I have used data classes before (and I prefer to use them!), but the reason why I have not used the dataclass decorator from [dataclasses](https://docs.python.org/3/library/dataclasses.html) in this case is because I am unsure what the RPM-based support story is for Juju and charmed operator. The EL7 and EL8 series of distributions only come with Python 3.6 out of the box, and dataclasses was not added to Python until version 3.8. This why I held off on using the dataclasses module. I had [similar issue](https://docs.python.org/3/library/dataclasses.html) in cleantest when I started adding support for more distributions.

If I can get charmed operators to work on a manually provisioned EL9 machine (they come with Python 3.9.14 out of the box), then I will look into using dataclasses instead. Until then I would prefer to not use the dataclasses module.

benhoyt replied:
We've just gone all in on Python 3.8+ as of `ops` 2.0 (see [release notes](https://github.com/canonical/operator/releases/tag/2.0.0)), as that's what is available on Focal. It'd be great if we can use EL9 if at all possible, so that we're on at least the same minimum Python version as `ops`. Plus, even if it takes a bit more figuring out, I think it's worth doing to get dataclasses and bunch of other more recent features.

NucciTheBoss replied:
Yeah, `ops` going to Python 3.8 is a current problem that I am working through with the folks who need to use CentOS. I am going to be looking into getting an EL9 base to work. It might be rather hacky at first but we'll get there :smile:

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+        (subprocess.CompletedProcess[str]):
+            returncode, stdout, and stderr of completed DNF command.
+    """
+    args = [args] if type(args) == str else args if type(args) == list else []
```
I guess I should warn you that I commonly use one-liners because I like their brevity. You can find them all throughout my personal projects :). I understand your point though so I will make it more readable for the general population.

benhoyt replied:
I think you'll find they're just as unreadable for you in 12 months time. :-) I've often written dense code like that in the past, then come back to it later and been like "who on earth wrote this code?" then done a `git blame` ... oops, that was me!

NucciTheBoss replied:
Removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+        """Determine if package is marked 'installed'.
+
+        Returns:
+            (bool): True if package is installed; False if otherwise.
```
Sounds good. Will double back to make this edit in all my docstrings.

NucciTheBoss replied:
Revised in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+
+    result = []
+    state = {"installed": _PackageState.INSTALLED, "available": _PackageState.AVAILABLE}
+    for pkg_name, pkg_version, pkg_repo in itertools.filterfalse(
+        lambda x: len(x) != 3,
+        [
+            line.strip("\n").strip().split()
+            for line in StringIO(
+                _dnf("list", postargs=[f"--{cache_type}"]).stdout.strip("\n"),
+            )
+        ],
+    ):
```
How does a comment block above this loop sound that walks through the steps? It might still be confusing even if I break it up slightly :sweat_smile:

Essentially I am:

1. Grabbing the dnf list of packages marked a certain may (installed or available).
2. Stripping the extra newline characters surrounding the captured output.
3. Wrapping the captured output with StringIO so I can iterate over the output like a file.
4. Strip newline characters, strip extra whitespace, and then split by whitespace.
5. Use `itertools.filterfalse(..., ...)` to remove splits that are not `len(...) == 3`. This gets rid of the "dnf metadata check" and "here are the types of packages I just" listed methods.

benhoyt replied:
I had the exact same comment as Jon below. I think part of the issue here is how the Black formatter is breaking things up (sometimes it's annoying like that), but part of it is the nested-ness of it, and the fact that the "main call", the `_dnf()` call, is kind of hidden away nested in the middle of things.

Comment blocks are useful when necessary, but I think here it'd be unnecessary if we wrote the simpler, non-nested version of the code (one of the versions I've shown below).

NucciTheBoss replied:
Dropped in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+        return self.__data.get("version", None)
+
+    @property
+    def fullversion(self) -> Optional[str]:
```
Sounds good. Will make more consistent.

NucciTheBoss replied:
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+        return self.__data.get("version", None)
+
+    @property
+    def fullversion(self) -> Optional[str]:
```
See my review comment here https://github.com/canonical/operator-libs-linux/pull/75#discussion_r1123291881

NucciTheBoss replied:
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on tox.ini:
```diff
@@ -92,4 +92,13 @@ description = Run integration tests
+        --ignore={[vars]tst_dir}integration/test_dnf.py \
+        --log-cli-level=INFO -s {posargs}
+
+[testenv:integration-but-with-cleantest]
```
I will change to a shorter name. I just added a separate tox env because I knew using cleantest for testing would be a topic for debate. I did not want to muck up the current integration-tests tox env while we hammer out the details behind how we should go about testing the dnf charm library.

NucciTheBoss replied:
Named changed to `integration-cleantest` in https://github.com/canonical/operator-libs-linux/pull/75/commits/5fea11961abc6efd241d8b171de3ca181f517513

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+
+
+class PackageInfo:
+    """Dataclass representing DNF package information."""
```
My comment here https://github.com/canonical/operator-libs-linux/pull/75#discussion_r1123193861 explains why I did not use a dataclass here. The TL;DR is that I do not confidently know what Enterprise Linux base I am going to be using. If I have to use EL8, dataclasses is off the table; if I can get EL9 to work; then I can use dataclasses.

NucciTheBoss replied:
Changed PackageInfo to a dataclass in https://github.com/canonical/operator-libs-linux/pull/75/commits/b55e859cb657ea618895133eacb7f883059f312c

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+    ):
+        placeholder = pkg_name.split(".")
+        epoch = None  # Not every DNF package has an epoch.
+        charset, buffer = deque(itertools.chain(pkg_version)), []
```
You are right; I can get away with just `deque(pkg_version)`. I just threw `itertools.chain(...)` here because I was like "iterable - I need this to be iterable" without really thinking that wait "str is iterable too!" :smile:

NucciTheBoss replied:
This style of parsing the package version string was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+    command: str,
+    args: Optional[Union[Union[str, os.PathLike], List[Union[str, os.PathLike]]]] = None,
+    preargs: Optional[List[str]] = None,
+    postargs: Optional[List[str]] = None,
```
I like this - I will incorporate it in!

NucciTheBoss replied:
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on tox.ini:
```diff
@@ -92,4 +92,13 @@ description = Run integration tests
-    pytest -v --tb native --ignore={[vars]tst_dir}unit --log-cli-level=INFO -s {posargs}
+    pytest -v --tb native --ignore={[vars]tst_dir}unit \
+        --ignore={[vars]tst_dir}integration/test_dnf.py \
+        --log-cli-level=INFO -s {posargs}
```
I can look into this (I agree that it makes the tox file much cleaner), but could a change like this potentially be out of scope for this pull request? I do not mind changing the other integration tests (and I would advocate for using cleantest), but I think this pull request should stay focused on getting the dnf charm library up to expectations. Perhaps a follow up pull request?

benhoyt replied:
I think considering it as out of scope for this PR is fair enough. If you could open an issue so we remember later that'd be great.

NucciTheBoss replied:
Opened a separate issue here https://github.com/canonical/operator-libs-linux/issues/76

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+            cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            universal_newlines=True,
```
Regarding `universal_newlines`, it is the same issue as I mentioned in https://github.com/canonical/operator-libs-linux/pull/75#discussion_r1123193861 that I do not have a set Python version yet. I need to check my information on this, but I believe that `text=True` did not become a kwarg until Python 3.7.

Good catch on the `check`! I will add that in.

NucciTheBoss replied:
Check added in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb, but I am still working on getting a newer EL base working.

NucciTheBoss replied:
Replaced `universal_newlines` with `text` in https://github.com/canonical/operator-libs-linux/pull/75/commits/d19923ec4e869150f7e257a6c55d8fcffffdaaf5

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+    """Dataclass representing DNF package information."""
+
+    def __init__(self, data: Dict[str, Union[str, _PackageState]]) -> None:
+        self.__data = data
```
Sounds good to me. I largely use the double __ when I want to make it a little more difficult to access a "private" variable. Given that our other code bases only use a single _, I will switch to that instead.

NucciTheBoss replied:
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+        if not hasattr(cls, f"{cls.__name__}__instance"):
+            cls.__instance = super(_Cache, cls).__new__(cls)
+            cls.__instance.__initialized = False
+        return cls.__instance
```
I can use that approach since `_Cache` is not meant to be directly accessed by end-users. I mostly did it this way because I was like "I only want one of these to exist at a time; ah I'll use the Singleton pattern!"

benhoyt replied:
Yeah, that's understandable. Often those "design patterns" come from the Java world, where you sort of need to do it differently. But I think in Python most people just use the module as a singleton. See for example https://stackoverflow.com/a/31887/68707

NucciTheBoss replied:
`_Cache` was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+                buffer.append(char)
+
+        result.append(
+            {
```
If I remember correctly, this was a change the formatter made after I wrote this block of code. I will look to see if I can change it back.

benhoyt replied:
Huh, you're right -- it's Black's fault...

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+                version = "".join(buffer)
+                buffer.clear()
+            else:
+                buffer.append(char)
```
I originally tried using regex to extract the values of epoch, version, and release, but they were not working the way I wanted to. The version regex would bork if there was no delimiter for epoch. I got annoyed at how much time this was taking me so I just quickly wrote the parser. I will revisit using the regex you mentioned here; I originally had three separate regexs I was using to extract epoch, version, and release.

NucciTheBoss replied:
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on tests/integration/test_dnf.py:
```diff
@@ -0,0 +1,170 @@
+        sys.exit(1)
+
+
+class TestDNF(unittest.TestCase):
```
I will add unit tests with mocking. I wrote the integration tests to prove it working :sweat_smile: 

[cleantest](https://nuccitheboss.github.io/cleantest/) is a custom testing framework that I have been working on for the past few months to make it easier to test code across multiple Linux distributions using LXD.

NucciTheBoss replied:
cleantest removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/8d1ccc56eeb08ce456f2749a2a9b0fbdd3a27314

NucciTheBoss commented on tests/integration/test_dnf.py:
```diff
@@ -0,0 +1,170 @@
+
+@lxd.target("test-ubuntu-jammy")
+def fail_on_ubuntu() -> None:
+    import sys
```
This has to do with cleantest. The body of this function is not actually run within the current Python process. It is instead picked up, injected into the specified LXD instance, and then a `Result` object is returned to the current process contained an exit code, stdout, and stderr. I call them [testlets](https://nuccitheboss.github.io/cleantest/user-guide/getting-started/#defining-a-testlet)!

NucciTheBoss replied:
cleantest removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/8d1ccc56eeb08ce456f2749a2a9b0fbdd3a27314

NucciTheBoss commented on tests/integration/test_dnf.py:
```diff
@@ -0,0 +1,170 @@
+        assert dnf["slurm-slurmrestd"].installed is True
+        sys.exit(0)
+    except (ImportError, AssertionError):
+        sys.exit(1)
```
Not exactly - this exit code inside the testlet is returned to the current Python process for evaluation in the `TestDNF` test case class. Simple binary way of determining if the testlet passed or failed. The `sys.exit(...)` here only affects the Python process run inside of the test environment instance.

NucciTheBoss replied:
cleantest removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/8d1ccc56eeb08ce456f2749a2a9b0fbdd3a27314

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+            )
+        ],
+    ):
+        placeholder = pkg_name.split(".")
```
I originally used placeholder because rpm package names can have multiple ".". Later in the block I just recombine the strings using `".".join(...)`. Seems like `.rsplit()` is exactly what I need here. I forget just how many built-ins Python has :sweat_smile:

NucciTheBoss replied:
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on tox.ini:
```diff
@@ -92,4 +92,13 @@ description = Run integration tests
+[testenv:integration-but-with-cleantest]
+description = Run integration tests using cleantest
+deps =
+    cleantest==0.4.0-rc1
```
I am not bringing pytest in as a dependency here because I am using the `unittest` module in this case. I did not want to super befuddle the tests before discussing the tests for the dnf charm library.

NucciTheBoss replied:
Opened issue https://github.com/canonical/operator-libs-linux/issues/76 so we can continue the discussion about testing charm libraries here.

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+            installed = executor.submit(_load_cache, "installed")
+            available = executor.submit(_load_cache, "available")
+            self.__installed = installed.result()
+            self.__available = available.result()
```
I need to think about this more :test_tube:

NucciTheBoss replied:
`_Cache` was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+    return result
+
+
+class _Cache:
```
I need to think about this more :test_tube:

NucciTheBoss replied:
`_Cache` was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+        """
+        return StringIO(_dnf("", preargs=["--version"]).stdout).readline().strip("\n")
+
+    def __getitem__(cls, package_name: str) -> PackageInfo:
```
`dnf info ...` can be used, however, the issue that I have with it is that its output is more difficult to parse (I would probably need to create a more complicated parser) than `dnf list ...` due to difference in how the returned information is delimited. The output of `dnf info ...` also changes depending on whether the package is installed, so I would need to also add handles for if those specific values are there or not.

Let me look into this :test_tube:

NucciTheBoss replied:
This was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb and replaced with the `fetch(...)` method.

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+        # Package cache can be large so perform searches in threads.
+        with ThreadPoolExecutor() as executor:
+            installed = executor.submit(_search, _Cache().installed, package_name)
+            available = executor.submit(_search, _Cache().available, package_name)
```
Let me play around with this :test_tube:

NucciTheBoss replied:
This was refactored out in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+    return None
+
+
+class _MetaDNF(type):
```
Hmm, I am conflicted about this. I have this check here because I do not want users to be able to import the dnf charm library if their base does not have DNF installed. The idea behind this is that they will have an implementation like the following:

```python3
try:
    from charms.operator_libs_linux.v0.dnf import dnf
    pkg_mngr = dnf
except ImportError:
    import charms.operator_libs_linux.v0.apt as apt
    pkg_mngr = apt
```

That is why I check for dnf when the class is first defined at import time (hence the metaclass). As for the `dnf install ...`, I have that there because otherwise the `add_repo(...)` method will not work due to the `config-manager` dnf subcommand not being available.

NucciTheBoss replied:
`_MetaDNF` was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+    @wraps(func)
+    def wrapper(*args, **kwargs) -> None:
+        func(*args, **kwargs)
+        _Cache().refresh()
```
I used the decorator so if I ever needed to change the refresh implementation, all I would have to do is modify the decorator. I can look into changing this however since we might want more granular controller over how the cache is refreshed.

NucciTheBoss replied:
The decorator was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+        )
+    except subprocess.CalledProcessError as e:
+        raise DNFExecutionError(
+            f"Failed to perform command {' '.join(cmd)}. Reason:\n\n{e.stderr}"
```
I found having the extra blank line made error messages easier to read, but I can remove it as it is more stylistic preference than anything else.

NucciTheBoss replied:
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented:
Thanks for reviewing this @jnsgruk and @benhoyt! I finished responding to most of your review comments. I am going to take dnf "back to the shop" and make some changes based on what has been discussed here. I will re-comment once I am done implementing my changes and ready for the next round of reviews.

jnsgruk commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+        return self.__data.get("repo", None)
+
+
+def _load_cache(
```
Yeh, I agree with Ben's comments below, we probably don't need to be threading things, given the context of where/how this code runs. The gains are likely marginal and probably not worth the complexity trade off.

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+
+        return (
+            f"{(f'{self.epoch}:' if self.epoch else '')}"
+            f"{self.version}{(f'-{self.release}' if self.release else '')}"
```
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+    ABSENT = "absent"
+
+
+class DNFExecutionError(Exception):
```
Revised in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+                _dnf("list", postargs=[f"--{cache_type}"]).stdout.strip("\n"),
+            )
+        ],
+    ):
```
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+        self.refresh()
+
+    @property
+    def installed(self) -> List[Dict[str, Union[str, _PackageState]]]:
```
This property was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+            (List[Dict[str, Union[str, _PackageState]]]):
+                Cache of available packages.
+        """
+        return copy.deepcopy(self.__available)
```
`_Cache` was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+            return PackageInfo({"name": package_name, "state": _PackageState.ABSENT})
+
+
+class dnf(metaclass=_MetaDNF):  # noqa N802
```
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+            *packages (str): Packages to upgrade on system.
+        """
+        if len(packages) == 0:
+            raise DNFExecutionError("No packages specified.")
```
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+    for pkg_name, pkg_version, pkg_repo in itertools.filterfalse(
+        lambda x: len(x) != 3,
+        [
+            line.strip("\n").strip().split()
```
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+    args: Optional[Union[Union[str, os.PathLike], List[Union[str, os.PathLike]]]] = None,
+    preargs: Optional[List[str]] = None,
+    postargs: Optional[List[str]] = None,
+) -> subprocess.CompletedProcess[str]:
```
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+#!/usr/bin/env python3
```
I add the `#!/usr/bin/env python3` hashbang to all my files so that I know what kind of file I am working with when reading outside of a Python-dedicated editor. I can remove it if we feel like it is necessary to be removed.

benhoyt replied:
Yeah, I think it sends the wrong signal. The hash bang says how to execute the script as a stand-alone executable (but it's not intended to be executable). Isn't the `.py` extension enough of a signal for "what kind of file" it is? The file extension is how editors normally auto-detect file type.

NucciTheBoss commented on tests/integration/test_dnf.py:
```diff
@@ -0,0 +1,170 @@
+        dnf.update()
+        dnf.install("epel-release")
+        dnf.install("slurm-slurmd", "slurm-slurmctld", "slurm-slurmdbd", "slurm-slurmrestd")
+        assert dnf["epel-release"].installed is True
```
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/6a2b799d76caf0eeb43519680c9f0ba435f5cd9b

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,462 @@
+        [
+            line.strip("\n").strip().split()
+            for line in StringIO(
+                _dnf("list", postargs=[f"--{cache_type}"]).stdout.strip("\n"),
```
`dnf list ...` output is more than likely intended to be human-readable, however, it does not have messages like `apt` about being an unstable interface. I am using its output because it is easy to parse with regexs, custom parsers, etc. dnf does have a `packages.db` file, but I have not figured out what the schema is.

dnf itself is written with Python (I do think dnf v5 is being written in C++ however), so it has utilities for interacting with `packages.db`, but the dnf Python package is only available on RPM-based distributions. I am avoidant towards using that library because this charm library will more than likely live inside charms that can also be deployed to Ubuntu bases.

NucciTheBoss commented:
Okay @jnsgruk and @benhoyt, I have completed my initial passthrough for improving the dnf charm library. Big changes that I made to this new revision is that I removed `_Cache` and `_MetaDNF` completely. Let me know what you think of the updated version. In the meantime, I will be testing an EL9 base with Juju to see if I can get it working. If it works, I will issue some new commits bumping up to Python 3.8 (i.e. change `PackageInfo` to a dataclass rather than having a lot of boilerplate).

benhoyt commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,239 @@
+        str: Captured stdout of executed DNF command.
+    """
+    if not installed():
+        raise ExecutionError(f"dnf not found on PATH {os.getenv('PATH')}")
```
Should we just let this fail? `subprocess.run` will raise `FileNotFoundError`, so if we did want to catch that explicitly and re-raise `ExecutionError` we could add another `except FileNotFoundError` case below. "It's easier to ask for forgiveness than permission." :-)

NucciTheBoss replied:
Let me try this :test_tube: - all I really want to do is ensure that dnf commands are not executed if dnf is not installed on the underlying system.

NucciTheBoss replied:
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/6518bcfbff7e0daa6ec871aba0ee14305f4a1cc8. Now `FileNotFoundError` is caught rather than performing a manual install check.

benhoyt commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,239 @@
+        PackageInfo: Information about package.
+    """
+    try:
+        status, info = _dnf("list", "-q", package).split("\n")[:2]  # Only take top two lines.
```
Nit: sometimes the "suffixes" like `.split()` and `[:2]` get hidden on the end of the lines. I found things like this somewhat clearer as:

```python
output = _dnf("list", "-q", package)
status, info = output.split("\n")[:2]  # Only take top two lines.
```

benhoyt reviewed (commented):
This is so much simpler, with any complexity that's left commensurate with what we're doing. Thanks! Just a few minor comments.

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,239 @@
+        *packages (str): Packages to upgrade on system.
+    """
+    if len(packages) == 0:
+        raise ExecutionError("No packages specified.")
```
Yeah, I caught this over the weekend while I was playing around with the library.

NucciTheBoss replied:
Fixed in https://github.com/canonical/operator-libs-linux/pull/75/commits/6518bcfbff7e0daa6ec871aba0ee14305f4a1cc8

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,239 @@
+    Args:
+        *packages (str): Packages to upgrade on system.
+    """
+    if len(packages) == 0:
```
The more you know! :smile:

NucciTheBoss replied:
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/6518bcfbff7e0daa6ec871aba0ee14305f4a1cc8

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,239 @@
+
+def update() -> None:
+    """Update all packages on the system."""
+    _dnf("update")
```
Yeah this was something I caught over the weekend as well :sweat_smile:. Here is how I changed it:

```python3
def upgrade(*packages: Optional[str]) -> None:
    """Upgrade one or more packages.

    Args:
        *packages (Optional[str]):
            Packages to upgrade on system. If None, upgrade all packages.
    """
    if len(packages) == 0:
        _dnf("upgrade")
    else:
        _dnf("upgrade", *packages)
```

_Ignore the fact that I am using `if len(packages) == 0`; I saw your comment later down in the thread._

Regarding `dnf upgrade ...` vs `dnf update ...`, having two separate functions for both `upgrade` and `update` is intentional due to something I am cooking up within the HPC team. I am trying to have parity between a custom APT charm library and this dnf charm library. Basically I am going for dnf and apt to have the same method signatures with little as possible difference between them so that it is easier to abstract them within an mixin/parent class. You can see my "prototype" (I am still working on the specification for supporting _alternative_ charm bases) [here](https://github.com/NucciTheBoss/cleantest/blob/0.4.0-dev/src/cleantest/utils/apt.py).

benhoyt replied:
I'm not at all sold on trying to make them conform to a standard interface (unless we were to do that formally with a typing.Protocol). But even so, I don't think it's a wise idea *here*. Package managers are all slightly different and use different terminology, and I think we'd be providing people with a "false sense of commonality" to kind of half give them parity. I think in `dnf.py` we should make the names/signatures DNF-like, and in `apt.py` we should make them APT-like.

If on top of that we need an abstracted-away interface, we can design that separately in a separate PR and/or charm lib. But even things like package names differ between package managers / repos, so I don't actually see that being very doable.

jnsgruk replied:
Yeh, agree. The user facing expectations of dnf and apt are really quite different, and these libraries should feel natural to people who know those tools - more a lightweight Python wrapper around each

NucciTheBoss replied:
Dropped the update method in https://github.com/canonical/operator-libs-linux/pull/75/commits/d19923ec4e869150f7e257a6c55d8fcffffdaaf5

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,239 @@
+    """
+    if len(packages) == 0:
+        raise TypeError("No packages specified.")
+    _dnf("remove", *packages)
```
My comment https://github.com/canonical/operator-libs-linux/pull/75#discussion_r1126639715 explains more. TL;DR is that I am going for parity between a dnf charm library and a custom apt charm library.

benhoyt replied:
See my response at https://github.com/canonical/operator-libs-linux/pull/75#discussion_r1127097918

NucciTheBoss replied:
Dropped the remove method in https://github.com/canonical/operator-libs-linux/pull/75/commits/d19923ec4e869150f7e257a6c55d8fcffffdaaf5

NucciTheBoss commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,239 @@
+        elif "Available" in status:
+            state = _PackageState.AVAILABLE
+        else:
+            state = _PackageState.UNKNOWN
```
> Oddly, in my Fedora VM, when I type `dnf install python2` it works, so it seems to be available. But when I do `dnf list -q python2` it displays "Error: No matching Packages to list" and returns a non-zero exit code. Any idea why, or what category of package `python2` might be? I'm just concerned we're missing a case here.

So I figured out the issue with `python2` on Fedora 36. `dnf install python2` works because dnf resolves the package name to `python2.7`; there is no `python2` package in the package archive, but there is a `python2.7` package that dnf uses instead. Since there is actually no `python2` package but a `python2.7` package, `dnf list -q python2` will say there is no package in existence while `dnf list -q python2.7` will work as expected.

NucciTheBoss replied:
Let me know what you think of this new resiliency check here:

https://github.com/canonical/operator-libs-linux/blob/6518bcfbff7e0daa6ec871aba0ee14305f4a1cc8/lib/charms/operator_libs_linux/v0/dnf.py#L169-L190

Rather than just using an _UNKNOWN_ state, I instead raise an error which tells the library to mark the package as absent. If there is faulty output, this means that something is wrong with the package and it should be marked as _ABSENT_. Other possible states are recent and obsolete, but these packages will by default either come up as _AVAILABLE_ or _INSTALLED_. Recent just means that the package was recently modified and obsolete means that there is just a newer version of the package available in the archive.

benhoyt replied:
> `dnf install python2` works because dnf resolves the package name to python2.7 ... Since there is actually no python2 package but a python2.7 package, `dnf list -q python2` will say there is no package

Is this going to be a problem with the design of using `dnf list` to get package info? I mean, the fact that `fetch` will return *ABSENT* but `install` can actually install it? Maybe at least worth mentioning in the doc comment.

benhoyt replied:
I was using Fedora 37 via `docker run -it fedora` BTW.

NucciTheBoss replied:
Sounds good. I replicated the same behavior in my Fedora 36 container via `lxc launch images:fedora/36 fedora-36-test`. I can add a note that the passed package name needs to exactly match the name of the package and cannot be an alias. The same issue is also present using `dnf info python2`:

```
[root@fedora-36-test ~]# dnf info python2
Last metadata expiration check: 2:11:58 ago on Mon Mar  6 20:26:31 2023.
Error: No matching Packages to list
```

NucciTheBoss replied:
Added note in https://github.com/canonical/operator-libs-linux/pull/75/commits/d19923ec4e869150f7e257a6c55d8fcffffdaaf5. Reading the `Notes:` section of the docstring will mention that the package name needs to be exact and not an alias.

jnsgruk commented:
Looking much, much simpler @NucciTheBoss - thanks! If you could clean up the unresolved conversations when you're done hacking we can take another pass at this and hopefully get it across the line?

benhoyt commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,241 @@
+    if not packages:
+        _dnf("upgrade")
+    else:
+        _dnf("upgrade", *packages)
```
I believe you don't need the `if` at all now and can just do `_dnf("upgrade", *packages)`, because Python doesn't mind passing an empty list as a `*args` list.

benhoyt replied:
That said, looking at what we have in the `install` and `remove` methods, we raise an explicit error if no packages were specified. So I'd recommend we have two methods here to avoid someone accidentally upgrading all packages when they meant "upgrade none":

```python
def upgrade_all():
    _dnf("upgrade")

def upgrade(*packages):
    if not packages:
        raise TypeError("No packages specified.")
    _dnf("upgrade", *packages)
```

NucciTheBoss replied:
Hmm... I need to think about this. Would folks be potentially confused if `dnf.upgrade()` did not upgrade the system and rather threw an error at them? Reason saying is that folks who know `dnf` are going to expect that `dnf.upgrade()` will upgrade the whole system if they pass no arguments. Provided [Jon's reasoning](https://github.com/canonical/operator-libs-linux/pull/75#discussion_r1127102180) about making the dnf charm library feel natural to those familiar with it, would `upgrade_all(...)` be a break from what is expected? Let me know your thoughts. Maybe we just include a note in the docstring the clearly calls out that omitting packages will upgrade the whole system.

benhoyt replied:
Yeah, that's fair. I don't have strong feelings either way -- making it follow `dnf upgrade` closely and documenting it seems very reasonable.

NucciTheBoss replied:
I decided to keep it as just upgrade in https://github.com/canonical/operator-libs-linux/pull/75/commits/d19923ec4e869150f7e257a6c55d8fcffffdaaf5. If people complain, we can add in a `upgrade_all()` method in a later pull request :sweat_smile:

NucciTheBoss commented:
Greetings @jnsgruk and @benhoyt - should be ready for the next round of reviews now. Here are the major changes you should be aware of now that I have published the new commits:

1. I removed the `update()` and `purge(...)` methods since they do not really exist in dnf.
2. I made `PackageInfo` a dataclass because I got CentOS 9 Stream to work as a base.
3. I removed cleantest from the integration tests and made some modifications to the current tox.ini file. TL;DR is that I also added an instance for CentOS 9 Stream to test dnf. See [this thread](https://chat.charmhub.io/charmhub/pl/k9w4xmu63pgjuja57edutk88sr) on the Charmhub Mattermost for more information. I also removed the integration test for checking if dnf is installed since I can just mock with `unittest.mock`.
4. I added unit tests. Coverage is at 100%, but I did have to add a `# pragma: no cover` to `add_repo(...)`. No matter how I patched the method, `coverage` would not mark it as covered. I think it had to do with trying to patch both `subprocess.run` and `dnf.fetch(...)`. Luckily the integration tests for the library validate that the `add_repo(...)` method works!

Let me know what you thoughts are. We are getting closer!

benhoyt commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,207 @@
+        # is probably in a bad state then.
+        version_match = re.match(r"(?:(.*):)?(.*)-(.*)", pkg_version)
+        if not version_match:
+            raise Error(f"Bad version {pkg_version}. Marking {package} absent.")
```
Shouldn't we just `return `PackageInfo(name=package)` here, instead of raising and re-catching the error? Also, might be better to be more explicit with the `state=ABSENT` field.

NucciTheBoss replied:
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/f75e25fb6c7def8537366ebc0256396eee54d342. No longer recapturing error and ABSENT `_state` assignment is explicit rather than implicit.

benhoyt reviewed (approved):
Nice. Approving, but just one nit comment!

jnsgruk commented on lib/charms/operator_libs_linux/v0/dnf.py:
```diff
@@ -0,0 +1,207 @@
+    """Install one or more packages.
+
+    Args:
+        *packages (Union[str, os.PathLike]): Packages to install on the system.
```
Optional nit: the Google styleguide suggests you can omit data types in your args/returns docstrings where type annotations are used - personally I think it makes things easier/less messy to read.

```suggestion
        *packages: Packages to install on the system.
```

jnsgruk replied:
This applies throughout :)

NucciTheBoss replied:
Sounds good. I removed the type annotations from the docstrings in https://github.com/canonical/operator-libs-linux/pull/75/commits/9f137e07fe136b2ab085de263ad7d5c647b8c05d. I just really like knowing what my expected types are :nerd_face:

jnsgruk reviewed (changes requested):
One final thing: we need to register the lib name and add LIBID / LIBAPI / LIBPATCH before this can be uploaded :)

NucciTheBoss commented:
_Working on adding module level documentation for how to use the dnf charm library for Charmhub now._

jnsgruk commented:
This is now published! https://charmhub.io/operator-libs-linux/libraries/dnf

Thanks @NucciTheBoss!

jamesbeedy commented:
@NucciTheBoss - super! Nice work, and thank you!
